CorsixTH/Src/th_lua_iso.cpp

```diff
@@ -22,38 +22,69 @@ SOFTWARE.
 
 #include "config.h"
 
+#include <stdexcept>
+#include <string_view>
+
 #include "iso_fs.h"
 #include "lua.hpp"
 #include "th_lua.h"
 #include "th_lua_internal.h"
 
 namespace {
 
+/**
+ * Lua binding to construct a new iso_filesystem object.
+ *
+ * Expects the lua stack to contain:
+ *   -# metatable for iso_fs
+ *   -# path to .iso file
+ *   -# (optional) path separator, defaults to '/'
+ *
+ * Returns on the lua stack:
+ *   -# userdata (iso_filesystem) or nil on failure
+ *   -# string (path) or nil on failure
+ *   -# error message if the first return value is nil
+ *
+ * @param L Lua stack
+ * @return Number of return values
+ */
 int l_isofs_new(lua_State* L) {
-  luaT_stdnew<iso_filesystem>(L, luaT_environindex, true);
-  return 1;
-}
+  const char* path = luaL_checkstring(L, 2);
+  std::string_view sep = luaL_optstring(L, 3, "/");
+  if (sep.length() != 1) {
+    luaL_argerror(L, 3, "Path separator must be a single character");
+  }
 
-int l_isofs_set_path_separator(lua_State* L) {
-  iso_filesystem* pSelf = luaT_testuserdata<iso_filesystem>(L);
-  pSelf->set_path_separator(luaL_checkstring(L, 2)[0]);
-  lua_settop(L, 1);
-  return 1;
+  try {
+    luaT_stdnew<iso_filesystem>(L, luaT_environindex, true, path, sep.at(0));
+  } catch (std::runtime_error& e) {
+    lua_pushnil(L);
+    lua_pushnil(L);
+    lua_pushfstring(L, "%s", e.what());
+    return 3;
+  }
+
+  // L
+  // 1 - table (metatable)
+  // 2 - path
+  // 3 - separator
+  // 4 - userdata (iso_filesystem)
+
+  lua_pushvalue(L, 2);
+  luaT_setenvfield(L, 4, "string");
+  lua_pushvalue(L, 2);
+  return 2;
 }
 
-int l_isofs_set_root(lua_State* L) {
-  iso_filesystem* pSelf = luaT_testuserdata<iso_filesystem>(L);
-  const char* path = luaL_checkstring(L, 2);
-  if (pSelf->initialise(path)) {
-    lua_pushvalue(L, 2);
-    luaT_setenvfield(L, 1, "string");
-    lua_settop(L, 1);
-    return 1;
-  } else {
-    lua_pushnil(L);
-    lua_pushstring(L, pSelf->get_error());
-    return 2;
+int l_isofs_is_valid_root(lua_State* L) {
+  // Static method so we don't need an instance
+  try {
+    iso_filesystem fs(luaL_checkstring(L, 1));
+    lua_pushboolean(L, true);
+  } catch (std::runtime_error& e) {
```

Comment:
Build generated warning here, do you need to fix? Will show up in all future PRs unless suppressed.